README.md
# ELnoK: A light front-end to Logstash/Elasticsearch

ELnoK stands for "Elasticsearch-Logstash-no Kibana". It was created to fulfil
some needs to inspect logs stored in Elasticsearch without having to rely on the
Kibana web interface.

It's a very simple tool to just output the logs lines corresponding to a specific
time period and/or strings matches in the fields. In practice, it can be handy if
you use Logstash and Elasticsearch to merge many logs together, and output them
synchronized, all in the same format.

## Environment
Requires Python 3.6+, and the "requests" Python package.
Tested only on Ubuntu 18.04, but probably will work on any Linux system, and could
even work on Windows.

## Installation
```
pip3 install -r requirements.txt
python3 setup.py build
sudo python3 setup.py install
```

Note, that alternatively, you can start elnok without installing it, using such lines:
```
export PYTHONPATH="src/"
python3 -m elnok ....
```

## Usage
On Linux, you can use the script `elnok` to run it. Note that it's inspired by
the journalctl command, so you might find familiar behaviour.

`elnok [OPTIONS...] [MATCHES....]`

### MATCHES
By default, it will list all the log. By specifying field matches, you can filter
the output so that the given fields of the log must contain at least the given
word. Matches are written in the format "field=value". Several matches can be
passed. In the case different fields are passed, all the matches will have to match.

### OPTIONS
* -h, --help

Show a help message and exit

* --version

Show program's version number and exit

* --log-level LEVEL

Set verbosity level (0-3, default = 0) of the elnok internals. 0 only shows errors.
A higher number shows more information about the program's behaviour.
Everything is logged on the error output. Unless you're debugging an issue
with elnok, you probably want to leave it to the minimum.

* --host HOST

Specify the name or IP address and port of the Elasticsearch server (default is localhost:9200).

* --index INDEX

Specify the index pattern to look into (default is logstash-*). It can be comma separated.

* --list-fields

List all the fields present in the index (tab separated)

* -o, --output OUTPUT

Controls the format of the generated output. Accepts "short" or "json".
The default is "short", which outputs each log on a line, tab/semicolon separated.
"json" shows each log line as a raw Elasticsearch hit

* --output-fields FIELDS

List of the fields to be printed (comma/semicolon separated).
On "json" output, by default all the fields are shown. On "short", the default
is  `"@timestamp,level,module,component,subcomponent:line,message"`. Commas are
replaced by tabs, and semicolons stay as is.
The fields available depend on the logstash storage. To find out the available ones,
use the json output, with the default behaviour.

* -S, --since SINCE, -U, --until UNTIL

Show entries within a limited period. SINCE is used to select the starting date
(by default is starts with the oldest log available). UNTIL is used to select
the last date shown. The date format follows Elasticsearch's format.
It can be either a full date like "2012-10-30T18:17:16", or can be a relative
date like "now-2d" or "now-1h". For a full description, see the
[Elasticsearch](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-daterange-aggregation.html)
[definitions](https://www.elastic.co/guide/en/elasticsearch/reference/current/common-options.html).

* -u, --username USERNAME

The Elasticsearch username (default is elastic).

* -p, --password PASSWORD

The Elasticsearch password (default is an empty string).

### Example

Shows all the log available:

        elnok

Store the log of 2 days ago into a file:

        elnok --log-level 1 -S now-2d/d -U now-1d/d > two-days-ago.log

Shows all log of the last 60 minutes with the ERROR level:

        elnok -S now-1h level=ERROR

Shows only the log where the field "subcomponent" is "img", and output a very short display:

        elnok -S now-5d -U now --output-fields @timestamp,level,message subcomponent=img

Shows the latest minutes logs in raw JSON format:

        elnok -S now-1m --output json


debian/changelog
elnok (1.1-4) focal; urgency=medium

  * renamed the Package from python3-elnok to elnok

 -- Nandish Patel <patel@delmic.com>  Tue, 27 Jun 2023 09:36:53 +0200

elnok (1.1-3) focal; urgency=medium

  * fix requests dependency in control Depends

 -- Nandish Patel <patel@delmic.com>  Wed, 21 Jun 2023 15:46:26 +0200

elnok (1.1-2) focal; urgency=medium

  * correct the copyright License statement

 -- Nandish Patel <patel@delmic.com>  Wed, 21 Jun 2023 15:02:26 +0200

elnok (1.1-1) focal; urgency=medium

  * Initial release for version 1.1

 -- Nandish Patel <patel@delmic.com>  Wed, 21 Jun 2023 14:29:59 +0200


src/elnok/es.py
# -*- coding: utf-8 -*-
'''
Created on 16 Apr 2021

@author: Éric Piel

Copyright © 2021 Éric Piel, Delmic

This file is part of ELnoK.

ELnoK is free software: you can redistribute it and/or modify it under the terms
of the GNU General Public License version 2 as published by the Free Software
Foundation.

ELnoK is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
ELnoK. If not, see http://www.gnu.org/licenses/.
'''
from collections import OrderedDict
import logging
import requests
from requests.auth import HTTPBasicAuth
from typing import Optional, Iterator, Dict, List, Set

# Elasticsearch API described here:
# https://www.elastic.co/guide/en/elasticsearch/reference/current/search-search.html
# Example call
# curl -X GET "localhost:9200/logstash-*/_search?pretty" -H 'Content-Type: application/json' -d'
#                                 {
#                                   "query": {
#                                     "match_all": {}
#                                   }
#                                 }
#                                 '
# Example output
# {
# :
#  "hits" : {
# :
#    "hits" : [
#      {
#        "_index" : "logstash-2021.04.12-000001",
#        "_type" : "_doc",
#        "_id" : "XlzGxngBpKXm13z6qcO-",
#        "_score" : 1.0,
#        "_source" : {
#          "message" : "SIM: parsing *ESR?",
#          "component" : "odemis",
#          "module" : "WCPCM",
#          "level" : "DEBUG",
#          "host" : "ericaspire",
#          "timestamp" : "2021-04-12 14:57:46,306",
#          "subcomponent" : "lakeshore",
#          "path" : "/var/log/odemis.log",
#          "@timestamp" : "2021-04-12T12:57:46.306Z",
#          "@version" : "1",
#          "line" : "494",
#          "raw_message" : "2021-04-12 14:57:46,306\tDEBUG\tlakeshore:494:\tSIM: parsing *ESR?"
#        }
#      },
#  :
# }
# SEARCH_URL =  "http://{host}/{target}/_search?"
SEARCH_MULTI_URL = "http://{host}/_search"
PIT_URL = "http://{host}/{target}/_pit"
MAPPING_URL = "http://{host}/{target}/_mapping"


def get_pit(host:str, target: str, username: str, password: str, keep_alive:float=60) -> dict:
    """
    keep_alive: how long the PIT should be valid (s)
    """

    url = PIT_URL.format(host=host, target=target)
    # Authentication
    auth = HTTPBasicAuth(username=username, password=password) if (username and password) else None
    response = requests.post(url, params={"keep_alive": "%ds" % keep_alive}, auth=auth)
    logging.debug(response.text)
    return response.json()["id"]


def search(host: str, target: str, username: str, password: str, match: Optional[Dict[str, str]]=None,
              since: Optional[str]=None, until: Optional[str]=None,
              fields: Optional[Set[str]]=None
              ) -> Iterator[OrderedDict]:
    """
    Does a elasticsearch query, by returning each hit one at a time via an iterator
    host: IP address/hostname + port of the elasticsearch server
    target: the name of the index. Can be a pattern.
    match: a mapping of field -> a filter on what to return. It follows the syntax of 
      ElasticSearch "match" query. See:
      https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-match-query.html
    since: filter for the minimum time
    until: filter for the maximum time. For the format. See:
      https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-daterange-aggregation.html
    fields: restrict the fields to return in the hit
    yield: OrderedDict (str -> value): each result (hit) found, in time ascending order
    """
    # TODO: add _source in query, to limit the fields returned
    # We don't receive a single "endless" response. Instead, we start a search,
    # and then "scroll" through it, by asking for more results. See:
    # https://www.elastic.co/guide/en/elasticsearch/reference/current/paginate-search-results.html

    # Get a "Point-in-time" (PIT), which is a sort of pointer to a snapshot of
    # the log, so that even if data changes, the paginated results don't change.
    pit = get_pit(host, target, username, password, keep_alive=10)

    # Keep requesting small amounts of data
    hits = None
    while True:
        url = SEARCH_MULTI_URL.format(host=host, target=target)
        req_data = {
            # Can be up to 10000. Any number "works", but too small cause a lot
            # of overhead, and too big causes latency.
            "size": 1000,
            "sort": [{"@timestamp": "asc"}],
            "pit": {"id": pit,
                    "keep_alive": "10s",  # Extend the PIT duration
            },
        }
        
        # Due to elasticsearch being a search engine, the query syntax is very
        # sophisticated. To look for multiple criteria (eg field match and time range),
        # it's necessary to combine them via a "bool,filter" query, which just
        # means all the criteria have to be fullfiled.
        # Note that it's fine to run a query with an empty filter.
        q_filters = []
        req_data["query"] = {
            "bool": {
                "filter": q_filters
            }
        }

        for field, val in match.items():
            # There are many types of text search "term" is looking for exactly the value
            # There might be better or more flexible options with match, wildcard or regexp...
            # https://www.elastic.co/guide/en/elasticsearch/reference/7.13/query-dsl-wildcard-query.html
            # q_filters.append({"wildcard": {field: {"value": val}}})
            # TODO: if the field is repeated, extend the query to work as a OR (ie, separate words with a space)
            q_filters.append({"match": {field: {"query": val}}})
            # TODO make the query case sensitive. It should be the matter of selecting
            # the right analyzer.
            # https://www.elastic.co/guide/en/elasticsearch/reference/current/specify-analyzer.html
            # However, for now adding "analyzer": "whitespace" seems to only make
            # the query text as-is, while the fields are always lowercase...

        # Add a time range, if requested. See:
        # https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-range-query.html
        q_time = {}
        if since:
            q_time["gte"] = since
        if until:
            q_time["lte"] = until

        if q_time:
            q_filters.append({"range": {"@timestamp": q_time}})

        # TODO: report properly if the date format is incorrect (ie, not understood by ES)

        if fields is not None:
            req_data["_source"] = list(fields)

        # Pass info from the previous request (if it's not the first one)
        if hits is not None:
            req_data["search_after"] = hits[-1]["sort"]

        # Authentication
        auth = HTTPBasicAuth(username=username, password=password) if (username and password) else None

        logging.debug("%s", req_data)
        response = requests.get(url, json=req_data, auth=auth)
        logging.debug(response.text)

        # Use OrderedDict in order to keep the order
        resp_dict = response.json(object_pairs_hook=OrderedDict)

        # In case there was an error parsing the query, it'll return "error" instead of "hits"
        if not "hits" in resp_dict:
            logging.error("Search query failed")
            if "error" in resp_dict:
                # TODO: make it prettier (it should be some kind of recursive text?
                for field, value in resp_dict["error"].items():
                    print("%s: %s" % (field, value))
            return

        hits = resp_dict["hits"]["hits"]
        if not hits:  # End of the search?
            return

        # Pass one each log line, one at a time
        for h in hits:
            yield h


def list_fields(host: str, target: str, username: str, password: str) -> List[str]:
    """
    List all the fields stored on the given index/indices
    host: IP address/hostname + port of the elasticsearch server
    target: the name of the index. Can be a pattern.
    """
    url = MAPPING_URL.format(host=host, target=target)
    # Authentication
    auth = HTTPBasicAuth(username=username, password=password) if (username and password) else None

    # See  https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-get-mapping.html
    response = requests.get(url, auth=auth)
    resp_dict = response.json()
    # Format:
    # list of str (index names) -> "mappings" -> "properties" -> dict str (field name) -> type
    fields: Set[str] = set()
    for idx_name, idx_desc in resp_dict.items():
        logging.debug("Parsing index %s", idx_name)
        try:
            fields_desc = idx_desc["mappings"]["properties"]
        except KeyError as ex:
            logging.info("Skipping index %s: %s", idx_name, ex)

        fields.update(fields_desc.keys())

    return list(fields)


src/elnok/output.py
# -*- coding: utf-8 -*-
'''
Created on 16 Apr 2021

@author: Éric Piel

Copyright © 2021 Éric Piel, Delmic

This file is part of ELnoK.

ELnoK is free software: you can redistribute it and/or modify it under the terms
of the GNU General Public License version 2 as published by the Free Software
Foundation.

ELnoK is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
ELnoK. If not, see http://www.gnu.org/licenses/.
'''
from datetime import datetime
import json
import logging
import string

TIME_FMT = "%Y-%m-%d %H:%M:%S.%f"
# OUTPUT_FMT = "{@timestamp}\t{level}\t{module}\t{component}\t{subcomponent}:{line}\t{message}"


class DefaultFormatter(string.Formatter):
    """
    String formatter which replaces missing keys by "∅" (empty symbol)
    """
    def get_value(self, key, args, kwargs):
        try:
            return super().get_value(key, args, kwargs)
        except KeyError:
            logging.info("Missing field %s", key)
            return "∅"


def print_hit(hit: dict, fmt: str):
    """
    Displays a hit (= one search result from ES) according to a given format
    hit: the elastic search response of the hit, as-is. It should contain a _source key.
    fmt: formatting string, with the fields to replace encoded as "{filed_name}"
    """
    source = hit["_source"]

    # Convert from timestamp -> epoch, and put it back into the user's format
    # example @timestamp: "2021-04-12T12:58:07.926Z"
    # TODO: do it lazyly in the formater
    if "@timestamp" in source:
        ts = datetime.strptime(source["@timestamp"], "%Y-%m-%dT%H:%M:%S.%fZ")
        source["@timestamp"] = ts.strftime(TIME_FMT)

    # If field missing => replace by empty symbol
    formatter = DefaultFormatter()
    try:
        print(formatter.format(fmt, **source))
    except KeyError:
        logging.exception("Failed to print %s", source)
        raise


def print_json_raw(hit: dict, fmt: str):
    """
    fmt: unused
    """
    source = hit["_source"]
    try:
        print(json.dumps(source))
    except KeyError:
        logging.exception("Failed to print %s", source)
        raise


src/elnok/__main__.py
# -*- coding: utf-8 -*-
'''
Created on 12 Apr 2021

@author: Éric Piel

Copyright © 2021 Éric Piel, Delmic

This file is part of ELnoK.

ELnoK is free software: you can redistribute it and/or modify it under the terms
of the GNU General Public License version 2 as published by the Free Software
Foundation.

ELnoK is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
ELnoK. If not, see http://www.gnu.org/licenses/.
'''

# Syntax is inspired by journalctl.
# Call like :
# elnok [OPTIONS...] [MATCHES...]
# -S, --since=, -U, --until=

import argparse
import logging
import re
import sys
from typing import Set

from elnok import es, output
import elnok

DEFAULT_OUTPUT_SHORT = "@timestamp,level,module,component,subcomponent:line,message"


def main(args: list) -> int:
    # arguments handling
    parser = argparse.ArgumentParser(description="A light front-end to Logstash/Elasticsearch")

    parser.add_argument('--version', dest="version", action='store_true',
                        help="Show program's version number and exit")
    parser.add_argument("--log-level", dest="loglev", metavar="<level>", type=int, choices=[0, 1, 2, 3],
                        default=0, help="Set verbosity level (0-3, default = 0) of the elnok internals")
    parser.add_argument("--host", default="localhost:9200",
                        help="Specify the name or IP address and port of the elasticsearch server (default is localhost:9200)")
    parser.add_argument("--index", default="logstash-*",
                        help="Specify the index pattern to look into (default is logstash-*). It can be comma separated.")
    parser.add_argument("--list-fields", dest="list", action='store_true',
                        help="List all the fields present in the index (tab separated)")
    parser.add_argument("--output", "-o", dest="output", default="short", choices=["short", "json"],
                        help="Controls the format of the generated output.\n "
                        "Default is short, which outputs each log on a line, tab/semicolon separated.\n "
                        "json shows each log line as a raw elasticsearch hit")
    parser.add_argument("--output-fields", dest="fields",
                        help="List of the fields to be printed (comma/semicolon separated)")
    parser.add_argument("--since", "-S", dest="since",
                        help="Show entries on or newer than the given date. Format is 2012-10-30T18:17:16 or now-2d.")
    parser.add_argument("--until", "-U", dest="until",
                        help="Show entries on or before the given date. Format is 2012-10-30T18:17:16 or now-1h.")
    parser.add_argument("matches", nargs="*",
                        help="Filter the output to only the fields that match")
    parser.add_argument("--username", "-u", type=str, default="elastic",
                        help="The Elasticsearch username (default is elastic).")
    parser.add_argument("--password", "-p", type=str, default="",
                        help="The Elasticsearch username's password (default is an empty string).")

    options = parser.parse_args(args[1:])

    # Cannot use the internal feature, because it doesn't support multiline
    if options.version:
        print(elnok.__shortname__ + " " + elnok.__version__ + "\n" +
              elnok.__copyright__ + "\n" +
              "Licensed under the " + elnok.__license__)
        return 0

    # Set up logging before everything else
    loglev_names = [logging.ERROR, logging.WARNING, logging.INFO, logging.DEBUG]
    loglev = loglev_names[options.loglev]

    # change the log format to be more descriptive
    logging.basicConfig(level=loglev, format='%(asctime)s (%(module)s) %(levelname)s: %(message)s')

    try:
        if options.list:
            fields_names = es.list_fields(options.host, options.index, options.username, options.password)
            print("\t".join(sorted(fields_names)))
            return 0

        # Convert matches from field=value to a dict field -> value
        matches = {}
        for m in options.matches:
            field, value = m.split("=")
            # TODO: support multiple times the same field (as a OR)
            if field in matches:
                raise ValueError("Cannot pass multiple matches on the same field (%s)" % (field,))
            matches[field] = value

        # Create the set of fields to retrieve
        if options.fields is None:
            # Pick different default fields based on output format
            if options.output == "short":
                options.fields = DEFAULT_OUTPUT_SHORT
            # For "json", we leave "None", which means all the fields

        fields = None  # all
        if options.fields:
            fields = set(f for f in re.split("[,:]", options.fields) if f)

        logging.debug("Selected fields are: %s", fields)

        # Create the formatting of the output, from the fields
        fields_fmt = None
        if options.fields:
            fields_fmt = []
            # Replace , -> \t and : stays :.
            # Each field is surrounded by {}.
            for m in re.finditer("[,:]+|[^,:]+", options.fields):
                p = m.group(0)
                # p is either a series of ,:, or a field name
                if not p:  # This shouldn't happen
                    logging.warning("Empty token %s", m)
                elif p[0] == ",":
                    fields_fmt.append("\t")
                elif p[0] == ":":
                    fields_fmt.append(":")
                else:
                    fields_fmt.append("{%s}" % (p,))

            fields_fmt = "".join(fields_fmt)

        logging.debug("Field format: %s", fields_fmt)

        if options.output == "short":
            print_output = output.print_hit
        elif options.output == "json":
            print_output = output.print_json_raw
        else:
            raise ValueError("Unknown output %s" % options.output)

        no_matches = True
        hit_fields: Set[str] = set()  # all the fields returned by the search
        for hit in es.search(options.host, options.index, options.username, options.password, match=matches, since=options.since, until=options.until, fields=fields):
            no_matches = False
            hit_fields.update(hit.get("_source", {}).keys())
            print_output(hit, fields_fmt)

        # If one of the fields to output was *never* returned, it might be that
        # there is a mistake in the field names to output. As it's easy to check,
        # we do it and report a error in this case.
        logging.debug("hit fields = %s, vs %s", hit_fields, fields)
        if fields:
            always_empty_fields = fields - hit_fields
            logging.debug("empty fields = %s", always_empty_fields)
            if always_empty_fields:
                fields_available = set(es.list_fields(options.host, options.index, options.username, options.password))
                wrong_fields = always_empty_fields - fields_available
                if wrong_fields:
                    logging.error("These fields do not exists: %s", ", ".join(sorted(wrong_fields)))
                    return 1

        # If nothing found, and the user has passed some matches, it could be that
        # there was a mistake in the field names of the matches. As it's easy to check, we do it and
        # report an "error" in this case.
        if no_matches and matches:
            # Something is strange => check if the user selected a field which doesn't exists
            fields_available = set(es.list_fields(options.host, options.index, options.username, options.password))
            wrong_fields = set(matches.keys()) - fields_available
            if wrong_fields:
                logging.error("These fields do not exists: %s", ", ".join(sorted(wrong_fields)))
                return 1

    except KeyboardInterrupt:  # Stopped by user
        logging.debug("Execution interrupted")
        return 128
    except Exception:
        logging.exception("Failure during execution")
        return 1

    return 0

ret = main(sys.argv)
exit(ret)

